### PR TITLE
Revert "Revert "Split prof gems into their own group"" 

### PR DESCRIPTION
**Pull request by Bo98**
Reland this so we can release a new portable Ruby, this time fixing `brew typecheck --update` to install all gems.

Obsoletes the fix in #15134 so we shouldn't need to reland that.

### CODE REVIEW

File: Library/Homebrew/utils/gems.rb [hunk 1]
```diff
@@ -12,6 +12,8 @@ module Homebrew
   # After updating this, run `brew vendor-gems --update=--bundler`.
   HOMEBREW_BUNDLER_VERSION = "2.3.26"
 
+  VALID_GEM_GROUPS = ["sorbet", "prof"].freeze
```
Comment (MikeMcQuaid) [inline Library/Homebrew/utils/gems.rb:15]: Could this be read from the `Gemfile` instead of being hardcoded?

File: Library/Homebrew/dev-cmd/typecheck.rb [hunk 2]
```diff
@@ -45,10 +45,11 @@ def self.typecheck_args
   def self.typecheck
     args = typecheck_args.parse
 
-    Homebrew.install_bundler_gems!(groups: ["sorbet"])
+    update = args.update? || args.update_all?
+    Homebrew.install_bundler_gems!(groups: update ? VALID_GEM_GROUPS : ["sorbet"])
```
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/typecheck.rb:49]: Could this get split out into a named variable? A bit hard to parse as-is.

File: Library/Homebrew/dev-cmd/vendor-gems.rb [hunk 3]
```diff
@@ -30,7 +30,7 @@ def vendor_gems
 
     Homebrew.install_bundler!
 
-    ENV["BUNDLE_WITH"] = "sorbet"
+    ENV["BUNDLE_WITH"] = VALID_GEM_GROUPS.join(":")
```
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/vendor-gems.rb:33]: Similarly: I think it'd be good that changes to this file always trigger the `vendor-gems` workflow. Perhaps even the `Gemfile` changing or `utils/gems` etc. should trigger this workflow too? Not really any harm in running it and would be good to catch at least the cases we had before where this only noticed to be failing after the fact.

File: Library/Homebrew/dev-cmd/typecheck.rb [hunk 4]
```diff
@@ -45,10 +45,11 @@ def self.typecheck_args
   def self.typecheck
```
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/typecheck.rb:45]: I think it'd be good that changes to this file always trigger the `sorbet` workflow.

**Review by MikeMcQuaid — COMMENTED** [review]
Thanks! Looking good so far.

File: Library/Homebrew/utils/gems.rb (hunk 1 above)
Comment (Bo98) [reply Library/Homebrew/utils/gems.rb:15]: Will have a look and see if there's a public API for that.

File: Library/Homebrew/dev-cmd/vendor-gems.rb (hunk 3 above)
Comment (Bo98) [reply Library/Homebrew/dev-cmd/vendor-gems.rb:33]: iirc it won't work on fork PRs as is but we could try add a "read-only" mode.

(same with Sorbet workflow below)

File: Library/Homebrew/utils/gems.rb (hunk 1 above)
Comment (MikeMcQuaid) [reply Library/Homebrew/utils/gems.rb:15]: We already use some `Gem` methods so should be doable.

File: Library/Homebrew/dev-cmd/vendor-gems.rb (hunk 3 above)
Comment (MikeMcQuaid) [reply Library/Homebrew/dev-cmd/vendor-gems.rb:33]: yeh, it won't commit to them but the failures we've seen have always occurred before committing so I think the "do it but don't push the result" flow would be worth having 👍🏻.

Comment (Bo98) [reply Library/Homebrew/dev-cmd/vendor-gems.rb:33]: Done in #15185 so I don't have to recreate this on a non-fork (but I'll rebase this on that before merging).

**Review by carlocab — APPROVED** [review]

**Review by MikeMcQuaid — APPROVED** [review]
Looks great thanks!
